### ISSUE TITLE
[YSQL] failed test: TestYsqlUpgrade#creatingSystemRelsAfterFailure

### ISSUE THREAD

**Issue by myang2021** [yugabyte/yugabyte-db#15216]
Jira Link: [DB-4396](https://yugabyte.atlassian.net/browse/DB-4396)
### Description

https://detective-gcp.dev.yugabyte.com/stability/test?class=org.yb.pgsql.TestYsqlUpgrade&name=creatingSystemRelsAfterFailure

Caused by https://phabricator.dev.yugabyte.com/D21198

**Comment by druzac** [comment]
Yep, this is d4b7fd4d346ac7e62d34921081cd3dd3c5c97e5f. That diff uses `emplace` to insert tables in the table index. On the create table path, we used to use:

```
(*table_ids_map_checkout)[table_id] = *table;
```

After commit, we use:
```
tables_.emplace(std::move(table));
```

There's a significant semantics difference when there are id collisions. The unit test `org.yb.pgsql.TestYsqlUpgrade.creatingSystemRelsAfterFailure` tries to create a system table twice. It uses the same table id for both attempts. Both before and after commit d4b7fd4d346ac7e62d34921081cd3dd3c5c97e5f there's a collision. The old code overwrote the table object in the map. The new code doesn't overwrite, keeping the existing entry and silently drops the new entry on the floor.

**Comment by myang2021** [comment]
Thanks for the update @druzac , the unit test is listed below

```
  @Test
  public void creatingSystemRelsAfterFailure() throws Exception {
    try (Connection conn = getConnectionBuilder().withDatabase(customDbName).connect();
         Statement stmt = conn.createStatement()) {
      setSystemRelsModificationGuc(stmt, true);

      String ddlSql = "CREATE TABLE pg_catalog.simple_system_table ("
          + "  id int"
          + ", CONSTRAINT simple_system_table_pk PRIMARY KEY (id ASC)"
          + "    WITH (table_oid = " + newSysOid() + ")"
          + ") WITH ("
          + "  table_oid = " + newSysOid()
          + ", row_type_oid = " + newSysOid()
          + ")";

      stmt.execute("SET yb_test_fail_next_ddl TO true");
      runInvalidQuery(stmt, ddlSql, "Failed DDL operation as requested");

      // Letting CatalogManagerBgTasks do the cleanup.
      Thread.sleep(BuildTypeUtil.adjustTimeout(5000));

      stmt.execute(ddlSql);
```
Note that the first attempt to create the system table failed due to error injection. Then the test waited for 5 seconds for the cleanup to happen. After that, the second attempt should succeed. Did we miss some cleanup code to remove the existing entry from the map?

**Comment by druzac** [comment]
@myang2021 there is no cleanup code to remove the entry from the map, as far as I can tell. Running this test on master before my commit (I was using f41220e9f54c9dddc48d8233dd0423e303226585), when the second creation attempt is processed in `CatalogManager::CreateTable`, there is a mapping for the id in `table_ids_map_`. We just replace it.

The only time mappings are removed from `table_ids_map_` is in `CatalogManager::AbortTableCreation`. We don't call this method on the `CleanupDeletedTables` path.

**Comment by myang2021** [comment]
@druzac got it. To get back the original behavior, is it possible to detect conflict (i.e., emplace failure) and then force an in-place overwrite?

**Comment by druzac** [comment]
Fixed by fdd3dc7b320128f80667a9742a0986177c69cf6b.
